### PR TITLE
Enable eslint for the build-tools project

### PR DESCRIPTION
**Pull request by tylerbutler**
Our build-tools are not using any lint config currently, so this change adds a baseline config using eslint/recommended and @typescript-eslint/recommended. I was motivated to do this because I want to use <https://typescript-eslint.io/rules/switch-exhaustiveness-check>

Code wise, this PR removes unneeded semicolons, uses const instead of let/var where possible, and removes unneeded escaping from regexes.

I didn't inherit from our shared config because doing so would require too many code changes to do in a single PR.

### CODE REVIEW

File: tools/build-tools/src/bumpVersion/context.ts [hunk 1]
```diff
@@ -98,6 +98,7 @@ export class Context {
         }
 
         const publishedPackageDependenciesPromises: Promise<void>[] = [];
+        // eslint-disable-next-line no-constant-condition
         while (true) {
```
Comment (ChumpChief) [inline tools/build-tools/src/bumpVersion/context.ts:102]: For a real fix, I think you could do something like
```ts
let pkg;
while ((pkg = pendingDepCheck.pop()) !== undefined) {
// ...
```

File: tools/build-tools/src/common/fileHashCache.ts [hunk 2]
```diff
@@ -25,4 +25,4 @@ export class FileHashCache {
     public clear() {
         this.fileHashCache.clear();
     }
-};
\ No newline at end of file
+}
```
Comment (ChumpChief) [inline tools/build-tools/src/common/fileHashCache.ts:28]: Nit: Is EOF newline not turned on here?

File: tools/build-tools/src/common/fluidUtils.ts [hunk 3]
```diff
@@ -52,6 +52,7 @@ async function inferRoot() {
             if (await isFluidRoot(curr)) {
```
Comment (ChumpChief) [inline tools/build-tools/src/common/fluidUtils.ts:52]: I think this whole try/catch/return could be replaced with something like
```ts
return isFluidRoot(curr).catch(() => false);
```

File: tools/build-tools/src/bumpVersion/context.ts (hunk 1 above)
Comment (ChumpChief) [reply tools/build-tools/src/bumpVersion/context.ts:102]: Same for a few other places I see - though some might take more thought than is worth right now

File: tools/build-tools/src/fluidBuild/tasks/leaf/miscTasks.ts [hunk 4]
```diff
@@ -144,13 +146,14 @@ export class GenVerTask extends LeafTask {
         try {
             const file = path.join(this.node.pkg.directory, "src/packageVersion.ts");
             const content = await readFileAsync(file, "utf8");
-            const match = content.match(/.*\nexport const pkgName = \"(.*)\";[\n\r]*export const pkgVersion = \"([0-9.]+)\";.*/m);
+            const match = content.match(/.*\nexport const pkgName = "(.*)";[\n\r]*export const pkgVersion = "([0-9.]+)";.*/m);
             return (match !== null && this.node.pkg.name === match[1] && this.node.pkg.version === match[2]);
+        // eslint-disable-next-line no-empty
         } catch {
         }
         return false;
```
Comment (ChumpChief) [inline tools/build-tools/src/fluidBuild/tasks/leaf/miscTasks.ts:154]:
```suggestion
        } catch {
            return false;
        }
```

File: tools/build-tools/src/repoPolicyCheck/handlers/assertShortCode.ts [hunk 5]
```diff
@@ -169,7 +170,7 @@ function writeShortCodeMappingFile() {
     }
 
     const fileContents =
-`/*!
+        `/*!
```
Comment (ChumpChief) [inline tools/build-tools/src/repoPolicyCheck/handlers/assertShortCode.ts:173]: This looks weird to me, maybe use a 1-line rule disable instead?

File: tools/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 6]
```diff
@@ -201,7 +201,7 @@ export const handlers: Handler[] = [
                 return { resolved: false, message: 'Error parsing JSON file: ' + file };
             }
 
-            let resolved = true;
+            const resolved = true;
```
Comment (ChumpChief) [inline tools/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:204]: Prob don't need this const at all, if it's just true.  Also -- looks like the prefer shorthand rule isn't working below in the return statement?

File: tools/build-tools/.eslintrc.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        "eslint:recommended",
+        "plugin:@typescript-eslint/recommended",
+    ],
+    "parserOptions": {
+        "project": ["./tsconfig.json", "./src/test/tsconfig.json"]
+    },
+    "rules": {
+        "@typescript-eslint/switch-exhaustiveness-check": "error",
+        "@typescript-eslint/no-inferrable-types": "off",
+        "@typescript-eslint/no-var-requires": "off",
```
Comment (ChumpChief) [inline tools/build-tools/.eslintrc.js:21]: I assume these are here because there were too many errors to practically do one-line disables?

**Review by ChumpChief — APPROVED** [review]

File: tools/build-tools/src/bumpVersion/context.ts (hunk 1 above)
Comment (tylerbutler) [reply tools/build-tools/src/bumpVersion/context.ts:102]: Yeah, I'm not trying to clean anything up right now -- just trying to enable the infrastructure so linting is possible. Follow up PRs can do more involved code changes.

File: tools/build-tools/src/common/fileHashCache.ts (hunk 2 above)
Comment (tylerbutler) [reply tools/build-tools/src/common/fileHashCache.ts:28]: No. Our shared config is not trivial to enable on a package that has never been linted, and it's not easy to adopt only the formatting rules. I've talked with @Josmithr about possible ways we could address this, but in the meantime, I want to be pragmatic and enable some baseline protection.

File: tools/build-tools/.eslintrc.js (hunk 7 above)
Comment (tylerbutler) [reply tools/build-tools/.eslintrc.js:21]: Yes, and across many files.

**Pull request merged**